File: surf-api-velocity/surf-api-velocity-server/src/main/java/dev/slne/surf/surfapi/velocity/server/impl/SurfVelocityApiImpl.java
package dev.slne.surf.surfapi.velocity.server.impl;

import static com.google.common.base.Preconditions.checkNotNull;

import com.velocitypowered.api.proxy.ProxyServer;
import dev.slne.surf.surfapi.core.server.impl.SurfCoreApiImpl;
import dev.slne.surf.surfapi.velocity.api.SurfVelocityApi;
import dev.slne.surf.surfapi.velocity.api.packet.SurfVelocityPacketApi;
import dev.slne.surf.surfapi.velocity.server.VelocityMain;
import dev.slne.surf.surfapi.velocity.server.impl.packet.SurfVelocityPacketApiImpl;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

/**
 * The SurfVelocityApiImpl class is an implementation of the SurfCoreApiImpl class. It provides
 * additional functionality specific to Surf Velocity.
 *
 * <p>
 * Example usage:
 * {@snippet :
 * import dev.slne.surf.surfapi.core.server.impl.SurfCoreApiImpl;
 * SurfCoreApiImpl<SurfVelocityPacketApi> surfApi = new SurfVelocityApiImpl();}
 * </p>
 */
@ApiStatus.Internal
public class SurfVelocityApiImpl extends SurfCoreApiImpl<SurfVelocityPacketApi> implements
    SurfVelocityApi {

  /**
   * The SurfVelocityApiImpl class is an implementation of the SurfCoreApiImpl class. It provides
   * additional functionality specific to Surf Velocity.
   * <p>
   * Example usage:
   * {@code  SurfCoreApiImpl<SurfVelocityPacketApi> surfApi = new SurfVelocityApiImpl();}
   */
  public SurfVelocityApiImpl() {
    super(new SurfVelocityPacketApiImpl());
  }

  @Override
  public ExecutorService getExecutorService() {
    return VelocityMain.getInstance().getExecutorService();
  }

  @Override
  public void sendPlayerToServer(UUID playerUuid, String server) {
    checkNotNull(playerUuid, "playerUuid");
    checkNotNull(server, "server");

    ProxyServer proxy = VelocityMain.getInstance().getServer();
    proxy.getPlayer(playerUuid)
        .ifPresent(player -> proxy.getServer(server).ifPresent(player::createConnectionRequest));
  }

  @Override
  public Optional<Object> getPlayer(@NotNull UUID playerUuid) {
    return Optional.ofNullable(
        VelocityMain.getInstance().getServer().getPlayer(playerUuid).orElse(null));
  }
}

File: surf-api-velocity/surf-api-velocity-server/src/main/kotlin/dev/slne/surf/surfapi/velocity/server/impl/SurfVelocityApiImpl.kt
package dev.slne.surf.surfapi.velocity.server.impl

import com.github.retrooper.packetevents.protocol.item.type.ItemType
import com.google.auto.service.AutoService
import com.velocitypowered.api.proxy.Player
import dev.slne.surf.surfapi.core.api.SurfCoreApi
import dev.slne.surf.surfapi.core.api.toast.Toast
import dev.slne.surf.surfapi.core.api.toast.ToastBuilder
import dev.slne.surf.surfapi.core.api.toast.ToastStyle
import dev.slne.surf.surfapi.core.api.util.checkInstantiationByServiceLoader
import dev.slne.surf.surfapi.core.server.impl.SurfCoreApiImpl
import dev.slne.surf.surfapi.core.server.impl.toast.ToastImpl
import dev.slne.surf.surfapi.velocity.api.SurfVelocityApi
import dev.slne.surf.surfapi.velocity.server.velocityMain
import net.kyori.adventure.text.Component
import java.util.*


@AutoService(SurfCoreApi::class)
class SurfVelocityApiImpl : SurfCoreApiImpl(), SurfVelocityApi {
    init {
        checkInstantiationByServiceLoader()
    }

    override val executorService get() = velocityMain.executorService
    override fun createToast(builder: ToastBuilder.() -> Unit): Toast =
        ToastBuilder().apply(builder).build()

    override fun createToast(
        icon: ItemType,
        text: Component,
        style: ToastStyle
    ) = ToastImpl(
        icon,
        text,
        style
    )

    override fun sendToast(
        player: Player,
        toast: Toast
    ) = toast.send(player.uniqueId)

    override fun sendPlayerToServer(playerUuid: UUID, server: String) {

        val proxy = velocityMain.server
        proxy.getPlayer(playerUuid).ifPresent { player ->
            proxy.getServer(server).ifPresent { server -> player.createConnectionRequest(server) }
        }
    }

    override fun getPlayer(playerUuid: UUID): Player =
        velocityMain.server.getPlayer(playerUuid).orElse(null)
}
